File: packages/packagechk/src/packagechk/claim.py
from __future__ import annotations

import json
import re
from dataclasses import dataclass
from pathlib import Path

MODULE_SEPARATOR_PATTERN = re.compile(r"[^A-Za-z0-9]+")


@dataclass(frozen=True)
class ClaimProjectSpec:
    package_name: str
    module_name: str
    description: str
    version: str


def module_name_from_package(package_name: str) -> str:
    module_name = MODULE_SEPARATOR_PATTERN.sub("_", package_name).strip("_").lower()
    if module_name == "":
        return "pkg"
    if module_name[0].isdigit():
        return f"pkg_{module_name}"
    return module_name


def render_claim_pyproject(spec: ClaimProjectSpec) -> str:
    package_path = f"src/{spec.module_name}"
    return f"""[build-system]
requires = ["hatchling"]
build-backend = "hatchling.build"

[project]
name = {_format_toml_string(spec.package_name)}
version = {_format_toml_string(spec.version)}
description = {_format_toml_string(spec.description)}
readme = {{text = "This package name is claimed.", content-type = "text/plain"}}
requires-python = ">=3.11"

[tool.hatch.build.targets.wheel]
packages = [{_format_toml_string(package_path)}]
"""


def render_claim_init_py(version: str) -> str:
    return f"""\"\"\"Claimed package name.\"\"\"

__version__ = {_format_toml_string(version)}
"""


def write_claim_project_files(project_dir: Path, spec: ClaimProjectSpec) -> None:
    module_dir = project_dir / "src" / spec.module_name
    if module_dir.exists():
        raise FileExistsError(f"module directory already exists: {module_dir}")

    module_dir.mkdir(parents=True)
    (project_dir / "pyproject.toml").write_text(render_claim_pyproject(spec), encoding="utf-8")
    (module_dir / "__init__.py").write_text(render_claim_init_py(spec.version), encoding="utf-8")


def _format_toml_string(value: str) -> str:
    return json.dumps(value, ensure_ascii=False)

File: ts/packages/packagechk/src/claim.ts
import { existsSync, mkdirSync, writeFileSync } from "node:fs";
import { join } from "node:path";

const MODULE_SEPARATOR_PATTERN = /[^A-Za-z0-9]+/g;

export interface ClaimProjectSpec {
	packageName: string;
	moduleName: string;
	description: string;
	version: string;
}

export interface NpmClaimProjectSpec {
	packageName: string;
	description: string;
	version: string;
	license: string;
}

export function moduleNameFromPackage(packageName: string): string {
	const moduleName = packageName
		.replaceAll(MODULE_SEPARATOR_PATTERN, "_")
		.replace(/^_+|_+$/g, "")
		.toLowerCase();
	if (moduleName === "") return "pkg";
	if (/^\d/.test(moduleName)) return `pkg_${moduleName}`;
	return moduleName;
}

export function renderClaimPyproject(spec: ClaimProjectSpec): string {
	const packagePath = `src/${spec.moduleName}`;
	return `[build-system]
requires = ["hatchling"]
build-backend = "hatchling.build"

[project]
name = ${formatTomlString(spec.packageName)}
version = ${formatTomlString(spec.version)}
description = ${formatTomlString(spec.description)}
readme = {text = "This package name is claimed.", content-type = "text/plain"}
requires-python = ">=3.11"

[tool.hatch.build.targets.wheel]
packages = [${formatTomlString(packagePath)}]
`;
}

export function renderClaimInitPy(version: string): string {
	return `"""Claimed package name."""

__version__ = ${formatTomlString(version)}
`;
}

export function writeClaimProjectFiles(projectDir: string, spec: ClaimProjectSpec): void {
	const moduleDir = join(projectDir, "src", spec.moduleName);
	if (existsSync(moduleDir)) throw new Error(`module directory already exists: ${moduleDir}`);
	mkdirSync(moduleDir, { recursive: true });
	writeFileSync(join(projectDir, "pyproject.toml"), renderClaimPyproject(spec), "utf8");
	writeFileSync(join(moduleDir, "__init__.py"), renderClaimInitPy(spec.version), "utf8");
}

export function renderNpmPackageJson(spec: NpmClaimProjectSpec): string {
	return `${JSON.stringify(
		{
			name: spec.packageName,
			version: spec.version,
			description: spec.description,
			license: spec.license,
			main: "index.js",
			files: ["README.md", "index.js"],
		},
		null,
		2,
	)}\n`;
}

export function renderNpmReadme(spec: NpmClaimProjectSpec): string {
	return `# ${spec.packageName}\n\nThis npm package name is claimed.\n`;
}

export function renderNpmIndexJs(): string {
	return "// Claimed package name placeholder.\n";
}

export function writeNpmClaimProjectFiles(projectDir: string, spec: NpmClaimProjectSpec): void {
	const packageJsonPath = join(projectDir, "package.json");
	if (existsSync(packageJsonPath))
		throw new Error(`package.json already exists: ${packageJsonPath}`);
	writeFileSync(packageJsonPath, renderNpmPackageJson(spec), "utf8");
	writeFileSync(join(projectDir, "README.md"), renderNpmReadme(spec), "utf8");
	writeFileSync(join(projectDir, "index.js"), renderNpmIndexJs(), "utf8");
}

function formatTomlString(value: string): string {
	return JSON.stringify(value);
}
